<a href="https://colab.research.google.com/github/lauquintero28/Business-Analitics/blob/main/1_NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Reto 1 Modelo Naive Bayes**

##0. Cargar las librerías.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np #Funciones matemáticas
import pandas as pd #Conexión con Archivos de Excel

from sklearn.naive_bayes import GaussianNB #Modelo Gaussian Naive Bayes
from sklearn.metrics import confusion_matrix #Metricas - Como lo voy a evaluar

##1. Cargar la base de datos.

In [ ]:
nxl="/content/drive/MyDrive/EAFIT/TERCER SEMESTRE/Analítica de Negocios/1. BD2_CarCrash_Int.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.dropna() #Evitar celdas vacias Not at Number
XDB.head()

#Creamos el DataMart
XD=XDB[["Weekday","ViolCat","ClearWeather","Month","Daylight"]] #Variables Entrada
yd=XDB[["Highway"]] #Valor Deseado - Aprenda el Modelo

## 2. Se procede con la configuración y evaluación del modelo Naive Bayes (Highway).

In [ ]:
mnb=GaussianNB()
mnb.fit(XD,yd)

#Analizamos los resultado arrojados por el modelo
u=mnb.theta_  #Son las medias
print("Las medias de las variables son: \n" ,u)

sigma=np.sqrt(mnb.var_) #Desviación Estandar
print("Los valores de las desviaciones son: \n" ,sigma)

#Se procede con el análisis de las variaciones de las variables
LS=u+sigma
print("Los limites superiores para las variables son: \n",LS)
LI=u-sigma
print("Los limites inferiores para las variables  son: \n",LI)

#Se procede con la evaluación del modelo
ydp=mnb.predict(XD)

#Se procede con la evaluación del comportamiento del modelo
cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es: \n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VP+VN)/len(XD)
print("La Exactitud del modelo es:",Ex)

Ter=(FP+FN)/len(XD)
print("La Tasa de error del modelo es:",Ter)

Sen=VP/(VP+FN)
print("La sensibilidad del modelo es:",Sen)

Esp=VN/(VN+FP)
print("La especificidad del modelo es:",Esp)

Pres=VP/(VP+FP)# precisión
print("La precisión del modelo es:",Pres)

PredNeg=VN/(VN+FN) #Predicción del negativo
print("La predicción Negativo del modelo es:",PredNeg)

#Para evaluar a la persona si tendrá o no un accidente
Persona=[[7,1,1,3,0]]
Evaluar=mnb.predict(Persona)
print("La persona tendrá un accidente:",Evaluar) #Si el modelo arroja '1' significa que 'SI' tendrá un accidente, si es '0' significa que 'NO' tendrá un accidente.

Las medias de las variables son: 
 [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
Los valores de las desviaciones son: 
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
Los limites superiores para las variables son: 
 [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
Los limites inferiores para las variables  son: 
 [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
La matriz de confusión es: 
 [[4560  325]
 [1563  257]]
La Exactitud del modelo es: 0.7184190902311708
La Tasa de error del modelo es: 0.2815809097688292
La sensibilidad del modelo es: 0.1412087912087912
La especificidad del modelo es: 0.9334698055271239
La precisión del modelo es: 0.4415807560137457
La predicción Negativo del modelo es: 0.7447329

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


##3. Se procede con la estimación de la predicción de la accidentalidad en autopista.

In [ ]:
XDBn=pd.read_excel(nxl,sheet_name=0)
XDBn.head(100)

#Creamos el DataMart
XDn=XDBn[['Weekday','ViolCat','ClearWeather','Month','Daylight']] #Variables Entrada
ydn=XDBn[['Highway']] #Valor Deseado - Aprenda el Modelo

#Hacer el pronóstico para todos los XDBs disponibles
ydpn=mnb.predict(XDn) #Hizo el pronóstico de la Preaprobación para los solicitantes

#Se procede con la creación del informe
XDsal=np.column_stack((XDn,ydpn,ydn)) #Para mostrar columnas juntas
df_sal=pd.DataFrame(XDsal)            #Se crea una tabla virtual
df_sal.to_excel('Informe.xlsx')       #Se exporta a Excel

#**Análisis de Resultados**

Los resultados arrojados por el modelo propuesto frente al pronóstico de Highway mostraron:

El modelo alcanzo una **Exactitud** del **72%** esto nos indica un buen desempeño en general, sin embargo, la alta tasa de error **28%** y la baja sensibilidad sugieren que el modelo podría tener problemas para detectar correctamente los casos positivos.

El modelo alcanzó una **Precisión** del **44%** esto nos indica que de todas las veces que el modelo predijo un accidente, el **44%** de las predicciones fueron correctas. Es decir, el modelo se equivoca en aproximadamente el **56%** de los casos en los que predice que ocurrirá un accidente.

El modelo alcanzó una **Sensibilidad** del **14%** esto nos indica que el modelo detecta correctamente solo el **14%** de los verdaderos casos de accidentes. Es decir, el modelo tiene dificultades para identificar la mayoría de los accidentes reales.

El modelo alcanzó una **Especificidad** del **93%** esto nos indica que el modelo identifica correctamente el **93%** de los casos en los que no ocurre un accidente. Es decir, el modelo es bastante bueno para predecir la ausencia de accidentes.

Es importante tener en cuenta que el modelo arroja un informe de resultados en donde se evalua la accidentalidad y se precide si un accidente automovilístico se provocó en una autopista o no. Como información adicional la base de datos posee un total de **6123** no accidentados, y un total de **582** accidentados.